### ISSUE TITLE
boolean indexing example on reference manual page triggers error in Numpy v1.15.3

### ISSUE THREAD

**Issue by XiangY-Q** [numpy/numpy#12289]
The boolean indexing with a boolean mask that has same rank ( or dimension) as array being indexed but different shape (has size 1 along certain dimension) triggers IndexError. The code below is exactly the same as sample code on the reference manual page, but instead of returning array([0, 1]), it triggers IndexError.
### Reproducing code example:

```python
import numpy as np
x = np.array([[0, 1], [1, 1], [2, 2]])
rowsum = x.sum(-1, keepdims=True)
x[rowsum <= 2]  # rowsum <= 2 has shape (3, 1) and (rows <= 2).nonzero() returns a a tuple of 2 shape (2,) integer ndarrays
```

### Error message:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: boolean index did not match indexed array along dimension 1; dimension is 2 but corresponding boolean dimension is 1

### Numpy/Python version information:
Python 3.6.5 (default, Apr 25 2018, 14:22:56) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
numpy Version: 1.15.3

**Comment by eric-wieser** [comment]
Same failure in 1.14.2

**Comment by seberg** [comment]
Just to note, the documentation needs to be adapted here. That example should probably be removed or replaced with a statement saying that the boolean index must have compatible shape.

**Comment by eric-wieser** [comment]
In 1.12.1, this gives:
```
VisibleDeprecationWarning: boolean index did not match indexed array along dimension 1; dimension is 2 but corresponding boolean dimension is 1
```

It also didn't do anything remotely reasonable anyway while it was supported

```python
>>> np.__version__
'1.12.1'
>>> x = np.arange(6).reshape(2, 3)
>>> x
array([[0, 1, 2],
       [3, 4, 5]])
>>> x[x.sum(axis=0, keepdims=True) < 10]
VisibleDeprecationWarning: ...
array([0, 1, 2])
>>> x[x.sum(axis=1, keepdims=True) < 10]
VisibleDeprecationWarning: ...
array([0])
```

**Comment by seberg** [comment]
Well, I guess broadcasting could be used. Not sure I like it much, but would have to think about it more. For `oindex` broadcasting is probably a bit more natural though (at least implementation wise, heh).

**Comment by bsipocz** [comment]
The documentation has been long fixed in #15694, and the example with the deprecated and now removed usage has been removed.
